modules/renter/contractor/downloader.go

```diff
@@ -51,25 +58,63 @@ func (hd *hostDownloader) Sector(root crypto.Hash) ([]byte, error) {
 
 // Close cleanly terminates the download loop with the host and closes the
 // connection.
-func (hd *hostDownloader) Close() error { return hd.downloader.Close() }
+func (hd *hostDownloader) Close() error {
+	hd.mu.Lock()
+	defer hd.mu.Unlock()
+	// Close is a no-op unless there is only one goroutine using the
+	// hostDownloader
+	hd.clients--
+	if hd.clients > 0 {
+		return nil
+	}
+	hd.contractor.mu.Lock()
+	delete(hd.contractor.downloaders, hd.contractID)
+	delete(hd.contractor.revising, hd.contractID)
+	hd.contractor.mu.Unlock()
+	return hd.downloader.Close()
+}
 
-// Downloader initiates the download request loop with a host, and returns a
-// Downloader.
-func (c *Contractor) Downloader(contract modules.RenterContract) (Downloader, error) {
+// Downloader returns a Downloader object that can be used to download sectors
+// from a host.
+func (c *Contractor) Downloader(id types.FileContractID) (Downloader, error) {
 	c.mu.RLock()
+	cachedDownloader, haveDownloader := c.downloaders[id]
 	height := c.blockHeight
+	contract, haveContract := c.contracts[id]
 	c.mu.RUnlock()
-	if height > contract.EndHeight() {
+	host, haveHost := c.hdb.Host(contract.NetAddress)
+
+	if haveDownloader {
+		// increment number of clients
+		cachedDownloader.mu.Lock()
+		cachedDownloader.clients++
+		cachedDownloader.mu.Unlock()
+		return cachedDownloader, nil
+	} else if !haveContract {
+		return nil, errors.New("no record of that contract")
+	} else if height > contract.EndHeight() {
```

Comment:
It's probably best to address this by adding a new function to the contract `RevisionLastCall` or something to that effect, so that the logic can be implemented in one place and used all over.